test/connection.test.ts

```diff
@@ -68,46 +68,272 @@ describe('Connection', function () {
 
-      await this.clientConn.end()
+      assert.calledOnce(connSpy.server.close)
+      assert.calledOnce(connSpy.server.end)
+      assert.calledOnce(connSpy.client.close)
+      assert.calledOnce(connSpy.client.end)
+
+      assert.calledTwice(serverStreamSpy.finish)
+      assert.calledTwice(serverStreamSpy.end) // Not being called twice
+      assert.calledTwice(serverStreamSpy.close)
+    })
 
-      assert.callCount(spy1, 1)
-      assert.callCount(spy2, 1)
+    it.skip('should remove the stream record once one side calls end', async function () {
+      // TODO figure out how to check the stream record was removed..
     })
 
-    it.skip('should close all incoming streams', async function () {
+    it('should complete sending all data from server when end is called on server side of the connection', async function() {
+      // TODO: Failing so skip for now
+      let data: Buffer[] = []
+      this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('data', (chunk: Buffer) => {
+          data.push(chunk)
+        })
+      })
+      const serverStream = this.serverConn.createStream()
+      serverStream.write(Buffer.alloc(30000))
+      await new Promise(setImmediate)
+      await this.serverConn.end()
+      assert.equal(Buffer.concat(data).length, 30000)
+    })
+
+    it('should complete sending all money from server when end is called on server side of the connection', async function () {
       const moneySpy = sinon.spy()
-      const dataSpy = sinon.spy()
+      let totalMoney: number = 0
       this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
-        stream.on('end', () => {
-          console.log('money stream end')
+        stream.setReceiveMax(12000)
+        stream.on('money', (amount) => {
           moneySpy()
+          totalMoney += +amount
         })
       })
-      this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
-        stream.on('finish', () => {
-          console.log('data strem end')
-          dataSpy()
+      const serverStream = this.serverConn.createStream()
+      serverStream.setSendMax(6000)
+      await this.serverConn.end()
+      assert.equal(totalMoney, 12000)
+      assert.callCount(moneySpy, 6)
+    })
+
+    it('should complete sending all data from client when end is called on client side of the connection', async function() {
+      // TODO: Failing so skip for now
+      let data: Buffer[] = []
+      this.serverConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('data', (chunk: Buffer) => {
+          data.push(chunk)
+        })
+      })
+      const clientStream = this.clientConn.createStream()
+      clientStream.write(Buffer.alloc(30000))
+      await new Promise(setImmediate)
+      await this.clientConn.end()
+      assert.equal(Buffer.concat(data).length, 30000)
+    })
+
+    it('should complete sending all money from client when end is called on client side of the connection', async function () {
```

Comment:
Theoretically the connection should work the same way whether it's the client or the server so I'm not sure we need to test the same behavior on both sides